packages/experience/src/hooks/use-prefilled-identifier.ts

```diff
@@ -0,0 +1,42 @@
+import { type SignInIdentifier } from '@logto/schemas';
+import { useContext, useMemo } from 'react';
+
+import UserInteractionContext from '@/Providers/UserInteractionContextProvider/UserInteractionContext';
+import { type IdentifierInputValue } from '@/components/InputFields/SmartInputField';
+
+import useLoginHint from './use-login-hint';
+
+type Options = {
+  enabledIdentifierType?: SignInIdentifier[];
+};
+
+const usePrefilledIdentifier = ({ enabledIdentifierType }: Options = {}) => {
+  const { identifierInputValue, getIdentifierInputValueByTypes } =
+    useContext(UserInteractionContext);
+
+  const loginHint = useLoginHint();
+
+  return useMemo<IdentifierInputValue>(() => {
+    // First, check if there's a cached input identifier
+    const cachedInputIdentifier = enabledIdentifierType
+      ? getIdentifierInputValueByTypes(enabledIdentifierType)
+      : identifierInputValue;
+
+    // If there's a cached input identifier, return it directly
+    if (cachedInputIdentifier) {
+      return cachedInputIdentifier;
+    }
+
+    // If there's no cached input identifier but a login hint exists, return the login hint
+    if (loginHint) {
+      return {
+        value: loginHint,
+      };
+    }
+
+    // If there's neither a cached input identifier nor a login hint, return empty
+    return { value: '' };
+  }, [enabledIdentifierType, getIdentifierInputValueByTypes, identifierInputValue, loginHint]);
+};
```

Comment:
I think this `useMemo` hook can be splitted into smaller parts, to avoid unnecessary re-calculation on dependencies change. For example, if there's cached input identifier, then whatever how the login hint changes, the return value is going to be the same. But whenever the login hint changes, the useMemo will be triggered for a re-calculation, which is unnecessary in such case.

Reply:
Good call, updated.